### PR TITLE
ENH: Add color parameter  to AnnotationBuilder.rectangle

### PR DESCRIPTION
**Pull request by hirenpopat67**
…ectangle function

### CODE REVIEW

**Comment by hirenpopat67** [comment]
I have added colour variable while draw rectangle on pdf change _annotations.py file rectangle function

**Comment by hirenpopat67** [comment]
> ## [Codecov](https://codecov.io/gh/py-pdf/pypdf/pull/1723?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=py-pdf) Report
> Patch coverage has no change and project coverage change: **`-0.03`** warning
> 
> > Comparison is base [(`9878034`)](https://codecov.io/gh/py-pdf/pypdf/commit/98780348eb6e819679e1dfb14c65d27cf6be9cc4?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=py-pdf) 92.50% compared to head [(`684e2ff`)](https://codecov.io/gh/py-pdf/pypdf/pull/1723?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=py-pdf) 92.47%.
> 
> Additional details and impacted files
> [umbrella View full report in Codecov by Sentry](https://codecov.io/gh/py-pdf/pypdf/pull/1723?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=py-pdf). loudspeaker Do you have feedback about the report comment? [Let us know in this issue](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=py-pdf).

I have added colour variable while draw rectangle on pdf change _annotations.py file rectangle function

**Comment by MartinThoma** [comment]
Thank you for the PR :pray: 

Should we rather call it `border_color`? 

Also, new parameters need to be added to the end for backwards compatibility. And I think we should start using keyword-only parameters, e.g.:

```
    @staticmethod
    def rectangle(
        rect: Union[RectangleObject, Tuple[float, float, float, float]],
        interiour_color: Optional[str] = None,
        *,  # This makes it keyword-only, see https://peps.python.org/pep-3102/
        rectangle_color: Optional[str] = None,  # I guess it's the border_color, but I didn't check.
    ) -> DictionaryObject:
```

**Comment by pubpub-zz** [comment]
As an alternative, as "/C" is used on many annotations I propose to add a function applicable to all annotations. example :
```
annotation = AnnotationBuilder.rectangle(rect=(50, 550, 200, 650)).set_border_color([255,0,0])
```
it would be great if the color could be either an arrayobject, a standard array or tuple of ints, and maybe an string from PIL.ImageColor.

@MartinThoma / @hirenpopat67  your opinion ?

**Comment by MartinThoma** [comment]
@pubpub-zz What would be the advantage of a method `set_border_color` vs a keyword-parameter `boder_color`?

**Comment by MartinThoma** [comment]
> it would be great if the color could be either an arrayobject, a standard array or tuple of ints, and maybe an string from PIL.ImageColor

Sounds good - but we should do it consistently for all color parameters in pypdf

**Comment by pubpub-zz** [comment]
> @pubpub-zz What would be the advantage of a method `set_border_color` vs a keyword-parameter `boder_color`?

the advantage would be to be able to have this solution for all annotations (circle, line,...)

also it will be a good approach later to add other fields/properties

**Comment by hirenpopat67** [comment]
*Dear @MartinThoma,*

Thank you for taking the time to review my pull request. I appreciate your
feedback and suggestions for improvement. I completely agree with your
comments and will make the necessary changes to ensure backwards
compatibility by making the argument a keyword-only one.

Additionally, I will rename the argument to something more descriptive such
as "border_color" to better reflect its purpose. These changes should help
make the code more readable and user-friendly.

Once again, thank you for your help and valuable feedback. I will work on
these changes and submit a revised pull request as soon as possible.

Best regards,

*Hiren Popat*

Software Engineer (Python Developer)

**Comment by MartinThoma** [comment]
> the advantage [of a method .set_border_color in contrast to a keyword-argument] would be to be able to have this solution for all annotations (circle, line,...)

You mean as a way to avoid code duplication?

The current annotation builder methods return DictionaryObjects. How would we change it so that this method can be added?

**Comment by pubpub-zz** [comment]
> > the advantage [of a method .set_border_color in contrast to a keyword-argument] would be to be able to have this solution for all annotations (circle, line,...)
> 
> You mean as a way to avoid code duplication?
> 
> The current annotation builder methods return DictionaryObjects. How would we change it so that this method can be added?

I would propose the AnnotationBuilder functions to create AnnotationDictionaryObject on which class the functions `set_border_color()` and later some other ones such as annotation_name("/NM"), border_type("/Border"),...

**Comment by hirenpopat67** [comment]
I have created the set_border_color() function and would like to request
your feedback before creating a new pull request to merge the code. Please
review the code and let me know if you have any suggestions or concerns.

Review:-
https://github.com/hirenpopat67/pypdf/blob/main/pypdf/generic/_annotations.py

Thank you for your attention to this matter. I look forward to hearing your
feedback.

Best regards,

*Hiren Popat*

Software Engineer (Python Developer)

**Comment by hirenpopat67** [comment]
As discussed i have created set_border_color function to avoid duplication

**Comment by hirenpopat67** [comment]
""

**Comment by hirenpopat67** [comment]
I hope this email finds you well. I am writing to express my gratitude for
your proposal to merge the code. I have reviewed the code and can confirm
that it is working flawlessly.

I am glad to hear that my contribution has helped in achieving the
project's goals.

Thank you once again..

Best regards,

*Hiren Popat*

Software Engineer (Python Developer)

**Comment by hirenpopat67** [comment]
> @hirenpopat67 I've pushed a PR ([hirenpopat67#1](https://github.com/hirenpopat67/pypdf/pull/1)) on your branch to show what I had in mind. What do you think of it ?

Great ,I have reviewed code its working fine.

Thank you for your time and consideration..

**Comment by hirenpopat67** [comment]
I hope this email finds you well. I am writing to express my gratitude for
your proposal to merge the code. I have reviewed the code and can confirm
that it is working flawlessly.

I am glad to hear that my contribution has helped in achieving the
project's goals.

Thank you once again..

Best regards,

*Hiren Popat*

Software Engineer (Python Developer)

**Comment by pubpub-zz** [comment]
@hirenpopat67 , can you complete test_generic.py in order to cover the new function

**Comment by MartinThoma** [comment]
I'm still not sure if the AnnotationObject / `set_border_color` really is a good solution. Do all annotations have a border. I know that /Rectangle is required, but for border I'm less certain.

Are there any other attributes that all annotations have?

**Comment by pubpub-zz** [comment]
from pdf 1.7 spec,
image

but it applicable to circle (border) strike out color, ...

maybe `set_color()` could be better...

File: pypdf/generic/_annotations.py [hunk 1]
```diff
@@ -12,6 +12,11 @@
 from ._rectangle import RectangleObject
 from ._utils import hex_to_rgb
 
+try:
+    from PIL import ImageColor as ImageColorLoaded
+except ImportError:
+    ImageColorLoaded = None
```
Comment (pubpub-zz) [inline pypdf/generic/_annotations.py:18]: to fix mypy
```suggestion
    ImageColorLoaded = None  # type: ignore
```

File: pypdf/generic/_annotations.py [hunk 2]
```diff
@@ -25,6 +30,28 @@ def _get_bounding_rectangle(vertices: List[Tuple[float, float]]) -> RectangleObj
...
+        Set the border color of a PDF annotation.
+
+        Args:
+            annotation: The PDF annotation to modify.
+            border_color: The color of the border in hex string format (e.g. "#FF0000" for red).
+        """
+        if isinstance(border_color, str):
+            if ImageColorLoaded is None:
+                raise ImportError("PIL required but not installed")
+            border_color = ImageColorLoaded.getrgb(border_color)
```
Comment (pubpub-zz) [inline pypdf/generic/_annotations.py:48]:
```suggestion
            border_color = (x/255.0 for x in ImageColorLoaded.getrgb(border_color)[:3])
```
for mypy

File: pypdf/generic/_annotations.py [hunk 3]
```diff
@@ -25,6 +30,28 @@ def _get_bounding_rectangle(vertices: List[Tuple[float, float]]) -> RectangleObj
...
+        if isinstance(border_color, (list, tuple)):
+            border_color = ArrayObject([FloatObject(n) for n in border_color[:3]])
```
Comment (pubpub-zz) [inline pypdf/generic/_annotations.py:50]:
```suggestion
            border_color = ArrayObject([FloatObject(n if n<=1.0 else n/255.0) for n in border_color[:3]])
```
if the array provided is an array of RGB color

**Review by pubpub-zz — CHANGES_REQUESTED** [review]
to fix mypy and also to cope with RGB on 0-255 ranges

**Comment by hirenpopat67** [comment]
I am writing to suggest a small improvement to the existing implementation of the rectangle function.

Currently, the function allows users to draw a rectangle on a PDF and specify the interior color of the rectangle using the **interior_color** argument. However, there is no option to specify the fill color of the rectangle.

Rather than create **set_border_colour** function we can create colour **color** keyword argument to the rectangle function
to make the function more user-friendly , I would like to suggest adding a "color" keyword argument to the rectangle function. This would allow users to specify both the border and fill color of the rectangle using a single function call.

Here is an example of the updated function signature:

@staticmethod
def rectangle(
rect: Union[RectangleObject, Tuple[float, float, float, float]],
interior_color: Optional[str] = None,
color: Optional[str] = None,
) -> AnnotationObject:


I believe that this change would make the function more intuitive and user-friendly. Please let me know if you have any questions or concerns regarding this suggestion.

Thank you for considering my suggestion.

Best regards,
*Hiren Popat*

Software Engineer (Python Developer)

**Comment by pubpub-zz** [comment]
> I am writing to suggest a small improvement to the existing implementation of the rectangle function.
> 
> Currently, the function allows users to draw a rectangle on a PDF and specify the interior color of the rectangle using the **interior_color** argument. However, there is no option to specify the fill color of the rectangle.
> 
> Rather than create **set_border_colour** function we can create colour **color** keyword argument to the rectangle function to make the function more user-friendly , I would like to suggest adding a "color" keyword argument to the rectangle function. This would allow users to specify both the border and fill color of the rectangle using a single function call.
> 

the approach I've proposed will be very efficient in this case adding the new function we will get:
```
AnnotationBuilder.rectangle((100,200,300,400),
     ).set_border_color("red",
     ).set_interior_color("yellow")
```
but without extra mods
```
AnnotationBuilder.ellipse((100,200,300,400),
     ).set_border_color("red",
     ).set_interior_color("yellow")
```

Do you agree ?

**Comment by hirenpopat67** [comment]
@pubpub-zz Yes, I agree. Your proposed approach of using the AnnotationBuilder class to create annotations,while setting the border and interior colors, is efficient and should work well in this case. The method chaining syntax you've used makes the code concise and easy to read.

**Comment by MartinThoma** [comment]
I think we are now having a way more complicated discussion within this PR. Let's move that to https://github.com/py-pdf/pypdf/discussions/1741 :-)

**Comment by MartinThoma** [comment]
Please wait with this PR until we come to a conclusion with https://github.com/py-pdf/pypdf/discussions/1741

It might be https://github.com/py-pdf/pypdf/pull/1745 - let's wait another week for opinions on that :-) (please also share your opinion about the #1745

**Comment by hirenpopat67** [comment]
> Please wait with this PR until we come to a conclusion with #1741
> 
> It might be #1745 - let's wait another week for opinions on that :-) (please also share your opinion about the #1745
@MartinThoma 
Thanks for asking my opinion i have shared #1741

**Pull request closed without merging**